### PR TITLE
[Hub Generated] Review request for Microsoft.HybridCompute to add version preview/2023-04-25-preview

### PR DESCRIPTION
**Pull request by vedkale**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/vedkale-hybridcompute-Microsoft.HybridCompute-2023-04-25-preview...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by pilor** [comment]
Please address automated checks and take this out of draft when you are ready for API review

**Review by raghushantha — APPROVED** [review]

**Comment by TimLovellSmith** [comment]
@vedkale the staging lint-diff validation errors/warnings for new apis, should be fixed as best practice:

https://github.com/Azure/azure-rest-api-specs/pull/23732/checks?check_run_id=13316840816

**Comment by TimLovellSmith** [comment]
>           },

resourcegroupname is in common-types too

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:284 in fb4f26d. [](commit_id = fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>             "name": "installPatchesInput",

nit, would suggest not using 'input' in the name.

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:356 in fb4f26d. [](commit_id = fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>             "name": "machineName",

nit, consider just defining machineNameParameter once in Parameters and $refing it

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:1311 in fb4f26d. [](commit_id = fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         }

my understanding is that it is always incorrect to use final-state-via with azure-async-operation, because final-state-via is supposed to explain where to get the final state of the RESOURCE, not the operation

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:1385 in fb4f26d. [](commit_id = fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2, deletion_comment = False)

**Comment by vedkale** [comment]
> > ```
> >         "name": "machineName",
> > ```
> 
> nit, consider just defining machineNameParameter once in Parameters and $refing it
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:1311 in [fb4f26d](https://github.com/Azure/azure-rest-api-specs/commit/fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2). [](commit_id = [fb4f26d](https://github.com/Azure/azure-rest-api-specs/commit/fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2), deletion_comment = False)

installPatches and assessPatches were added in [this](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2022-12-27) api version but left out in preview version, so we just copied it from previous stable version for consistency

**Comment by vedkale** [comment]
> > ```
> >     }
> > ```
> 
> my understanding is that it is always incorrect to use final-state-via with azure-async-operation, because final-state-via is supposed to explain where to get the final state of the RESOURCE, not the operation
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:1385 in [fb4f26d](https://github.com/Azure/azure-rest-api-specs/commit/fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2). [](commit_id = [fb4f26d](https://github.com/Azure/azure-rest-api-specs/commit/fb4f26d8eff5951aa046bfe6bbd7e9af97c312b2), deletion_comment = False)

@TimLovellSmith changing final-state-via to location makes the PR fail the linter check and causes breaking changes as it is inconsistent with the previous API versions

**Comment by vedkale** [comment]
> @vedkale the staging lint-diff validation errors/warnings for new apis, should be fixed as best practice:
> 
> https://github.com/Azure/azure-rest-api-specs/pull/23732/checks?check_run_id=13316840816

@TimLovellSmith Looks like the ResourceMustReferenceCommonTypes and LROPostFinalStateViaProperty are being removed/fixed, would it be better to change final-state-via to location or to keep it as azure-async-operation to remain consistent with old API versions?

**Comment by TimLovellSmith** [comment]
@ArcturusZhang can you please confirm my understanding whether they should not use final-state-via, and the 'breaking' change would actually be a bugfix?

**Comment by TimLovellSmith** [comment]
@vedkale the linter rule may be being fixed somehow, but inheriting ProxyResource or TrackedResource rather than Resource or using the x-ms-resource annotation directly, is still ordinarily best practice. Do neither TrackedResource nor ProxyResource apply in your scenario for some reason?

**Comment by TimLovellSmith** [comment]
>           "$ref": "../../../../../common-types/resource-management/v5/types.json#/definitions/TrackedResource"

Why did you want MachineRunCommand to be tracked instead of proxy by the way?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3098 in b7db538. [](commit_id = b7db538ee0b5855f283f013aef8165c7d3440788, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>       "x-ms-azure-resource": true

I don't think you really need these parts, since they're implied by inheriting from Resource.

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:2611 in b7db538. [](commit_id = b7db538ee0b5855f283f013aef8165c7d3440788, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
@vedkale if its a breaking change to fix final-state-via for existing resource types, we should try to decide whether its a desirable one (and follow the breaking change process) or an undesirable one. But for the new resource types, I don't think we need worry about breaking changes, and can just focus on correctness.

**Comment by vedkale** [comment]
> > ```
> >       "$ref": "../../../../../common-types/resource-management/v5/types.json#/definitions/TrackedResource"
> > ```
> 
> Why did you want MachineRunCommand to be tracked instead of proxy by the way?
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3098 in [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788). [](commit_id = [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788), deletion_comment = False)

MachineRunCommand is supposed to mirror the VirtualMachineRunCommand from Microsoft.Compute 

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-03-01/runCommand.json

**Comment by vedkale** [comment]
> > ```
> >   "x-ms-azure-resource": true
> > ```
> 
> I don't think you really need these parts, since they're implied by inheriting from Resource.
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:2611 in [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788). [](commit_id = [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788), deletion_comment = False)

This was also inherited from older API version

**Comment by vedkale** [comment]
> @vedkale if its a breaking change to fix final-state-via for existing resource types, we should try to decide whether its a desirable one (and follow the breaking change process) or an undesirable one. But for the new resource types, I don't think we need worry about breaking changes, and can just focus on correctness.

@TimLovellSmith reverted the change to fix issue across versions and consistency, it doesn't seem like a big issue and we would rather have the APIs be consistent

**Comment by TimLovellSmith** [comment]
I think the definition of 'Resource' they are using is mostly equivalent to 'TrackedResource' in the common-types types.json intended for use by all services.
In general we'd rather everyone were using that 'TrackedResource' definition, since its clearer in intent, to reviewers that you intend to be a tracked resource with location + tags, and we don't have to worry that the multiple definitions could be inconsistent.
Its just considered best practice rather than mandatory for now, so consider it optional.

**Comment by vedkale** [comment]
@TimLovellSmith if everything looks good, can we get an approval so we can start rolling out the manifest?

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_ListBySubscription.json [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+              "displayName": null,
+              "machineFqdn": null,
+              "vmUuid": null,
+              "osSku": null,
+              "domainName": null,
+              "adFqdn": null,
+              "dnsFqdn": null,
+              "osVersion": null,
+              "osType": null,
+              "osProfile": {
```
Comment (TimLovellSmith) [inline specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_ListBySubscription.json:37]: The examples look incomplete. Are these properties actually supposed to be nullable?

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_CreateOrUpdate.json [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                "patchMode": null
+              }
+            },
+            "linuxConfiguration": {
+              "patchSettings": {
+                "assessmentMode": null,
+                "patchMode": null
+              }
+            }
+          },
```
Comment (TimLovellSmith) [inline specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_CreateOrUpdate.json:66]: Are there supposed to be this many nulls?

**Comment by TimLovellSmith** [comment]
>     "WindowsParameters": {

should the defnitions be renamed WindowsPatchParameters / LinuxPatchParameters to avoid confusing with other windows or linux parameters in other requests?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:2862 in 18195f9. [](commit_id = 18195f9ad592cdbed1ff56af9078bfa7152371ce, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "$ref": "../../../../../common-types/resource-management/v5/types.json#/definitions/TrackedResource"

API consistency? Tagging?

---
In reply to: [1540748829](https://github.com/Azure/azure-rest-api-specs/pull/23732#issuecomment-1540748829) [](http://example.com/codeflow?ancestors=1540748829,1540748829)

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3098 in b7db538. [](commit_id = b7db538ee0b5855f283f013aef8165c7d3440788, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "The provisioning state, which only appears in the response."

Why isn't this provisioningState an enum with known values?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3167 in 18195f9. [](commit_id = 18195f9ad592cdbed1ff56af9078bfa7152371ce, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         },

Are these secret uris with SAS signatures? Or non-secret, unsigned uris?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3155 in 18195f9. [](commit_id = 18195f9ad592cdbed1ff56af9078bfa7152371ce, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         },

Its secret and intended to be passed only by the customer I assume.
Add x-ms-secret. Should it be write-only, to avoid information disclosure?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3142 in 18195f9. [](commit_id = 18195f9ad592cdbed1ff56af9078bfa7152371ce, deletion_comment = False)

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_ListBySubscription.json (hunk 1 above)
Comment (vedkale) [reply specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_ListBySubscription.json:37]: Yes, a lot of these properties are filled in specific scenarios, for example the kind property is filled in when the machine is onboarded in a private cloud scenario, the version of the agent installed on the machine also affects which properties get filled

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_CreateOrUpdate.json (hunk 2 above)
Comment (vedkale) [reply specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/examples/Machines_CreateOrUpdate.json:66]: Yes, see previous comment

**Comment by vedkale** [comment]
> > ```
> > "WindowsParameters": {
> > ```
> 
> should the defnitions be renamed WindowsPatchParameters / LinuxPatchParameters to avoid confusing with other windows or linux parameters in other requests?
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:2862 in [18195f9](https://github.com/Azure/azure-rest-api-specs/commit/18195f9ad592cdbed1ff56af9078bfa7152371ce). [](commit_id = [18195f9](https://github.com/Azure/azure-rest-api-specs/commit/18195f9ad592cdbed1ff56af9078bfa7152371ce), deletion_comment = False)

We could do that, but doing so would cause breaking changes against old API versions

**Comment by vedkale** [comment]
> > ```
> >       "$ref": "../../../../../common-types/resource-management/v5/types.json#/definitions/TrackedResource"
> > ```
> 
> API consistency? Tagging?
> 
> In reply to: [1540748829](https://github.com/Azure/azure-rest-api-specs/pull/23732#issuecomment-1540748829)
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3098 in [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788). [](commit_id = [b7db538](https://github.com/Azure/azure-rest-api-specs/commit/b7db538ee0b5855f283f013aef8165c7d3440788), deletion_comment = False)

Main reason is to make sure the API is consistent with Microsoft.Compute's version of RunCommand

**Comment by vedkale** [comment]
> > ```
> >     },
> > ```
> 
> Are these secret uris with SAS signatures? Or non-secret, unsigned uris? If they're secret, annotate with x-ms-secret. They're intended to be provided by the customer I assume... in that case should they be write-only?
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3155 in [2fb299e](https://github.com/Azure/azure-rest-api-specs/commit/2fb299ee3c5d3efa9a06354dc51dc07ad7421436). [](commit_id = [2fb299e](https://github.com/Azure/azure-rest-api-specs/commit/2fb299ee3c5d3efa9a06354dc51dc07ad7421436), deletion_comment = False)

They would be just the uris and the tokens would be passed in the protected parameters

**Comment by vedkale** [comment]
> > ```
> >     },
> > ```
> 
> Its secret and intended to be passed only by the customer I assume. Add x-ms-secret. Should it not be write-only, to avoid information disclosure? (x-ms-mutability create/update)
> 
> Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3142 in [2fb299e](https://github.com/Azure/azure-rest-api-specs/commit/2fb299ee3c5d3efa9a06354dc51dc07ad7421436). [](commit_id = [2fb299e](https://github.com/Azure/azure-rest-api-specs/commit/2fb299ee3c5d3efa9a06354dc51dc07ad7421436), deletion_comment = False)

added x-ms-secret

**Comment by TimLovellSmith** [comment]
>         },

x-ms-mutability: "create, update"?

---
Refers to: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3143 in c9a4b57. [](commit_id = c9a4b57465e896b6c8cd13c53e5ddfe4441868e4, deletion_comment = False)

**Comment by justinytchen** [comment]
image
Looks good, small differences in descriptions of properties compared to Microsoft.Compute, but I'm not sure if these have been accounted for yet.

**Comment by vedkale** [comment]
@Wzb123456789 can I please get a review for this PR? It looks like CredScan is failing but there are no secrets in the example, is that ok?

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json [hunk 3]
```diff
@@ -0,0 +1,3315 @@
...
+    }
+  },
+  "parameters": {
+    "machineNameParameter": {
+      "name": "machineName",
+      "in": "path",
+      "required": true,
+      "type": "string",
+      "pattern": "[a-zA-Z0-9-_\\.]",
+      "description": "The name of the machine containing the run command."
```
Comment (qiaozha) [inline specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json:3312]: please add x-ms-parameter-location: method for this parameter. Thanks

Comment (vedkale) [reply specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-04-25-preview/HybridCompute.json]: @qiaozha done

**Comment by vedkale** [comment]
@Wzb123456789 can I please get a review for this PR?

**Comment by raych1** [comment]
Go sdk breaking change is false alarm.

**Comment by raghushantha** [comment]
Awesome! Thank you @***@***.***>, @Crystal ***@***.***>

@***@***.***> as fyi
________________________________

**Pull request merged**